# EXPLORATION 02
## | 프로젝트 (2) load_wine 와인 품종 분류
### | 프로젝트 학습 과정
1. 필요한 모듈 import 하기
2. 데이터 준비 및 데이터 이해
- Feature Data 지정하기
- Label Data 지정하기
- Target Names 출력해 보기
- 데이터 Describe 해 보기
3. train, test 데이터 분리
4. 다양한 모델로 학습 시켜보기
- Decision Tree 사용해 보기
- Random Forest 사용해 보기
- SVM 사용해 보기
- SGD Classifier 사용해 보기
- logistic Regression 사용해 보기
5. 모델 평가
- sklearn.metrics에서 제공하는 평가지표 중 적절한 것 선택
- 선택 이유

# 2. load_wine  와인 분류

### Data info
1. Classes: 3
2. Samples per class: [59,71,48]
3. Samples total: 178
4. Dimensionality: 13
5. Features: real, positive

## 2_1. 필요한 모듈 import

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
%matplotlib inline

from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

## 2_2. 데이터 준비 및 이해

In [2]:
wine = load_wine()
wine_data = wine.data # x_data
wine_label = wine.target # y_data
wine_data.shape

(178, 13)

In [3]:
wine_label.shape

(178,)

In [4]:
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [5]:
wine.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [6]:
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [7]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [8]:
wine.target[[10, 80, 140]]

array([0, 1, 2])

In [9]:
wine_data[0]

array([1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
       3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
       1.065e+03])

In [10]:
df = pd.DataFrame(wine_data, columns=wine.feature_names)
df.columns
df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [11]:
df.shape

(178, 13)

In [12]:
df.describe()

alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       color_intensity         hue  od280/od315_of_diluted_wines      proline  
count       178.000000  178.000000                    178.000000   178.000000  
mean          5.058090    0.957449                      2.611685   746.893258  
std           2.318286    0.228572                      0.709990   314.907474  
min           1.280000    0.480000                      1.270000   278.000000  
25%           3.220000    0.782500                      1.937500   500.500000  
50%           4.690000    0.965000                      2.780000   673.500000  
75%           6.200000    1.120000                      3.170000   985.000000  
max          13.000000    1.710000                      4.000000  1680.000000

## 2_3. train, test 데이터 분리

In [13]:
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size=0.3,
                                                   random_state=32)

## 2_4. 모델 학습
### 2_4_(1) Decision Tree

In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

decision_tree = DecisionTreeClassifier(random_state=32)
print(decision_tree._estimator_type)

classifier


In [15]:
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=32)

In [16]:
Decision_pred = decision_tree.predict(X_test)
Decision_pred

array([1, 1, 0, 2, 2, 0, 0, 2, 0, 0, 1, 2, 0, 0, 0, 2, 1, 1, 0, 0, 1, 2,
       2, 2, 1, 1, 0, 0, 0, 0, 1, 0, 2, 1, 0, 1, 2, 0, 0, 2, 0, 1, 2, 2,
       0, 0, 0, 2, 0, 1, 1, 1, 1, 1])

In [17]:
accuracy = accuracy_score(y_test, Decision_pred)
accuracy

0.9629629629629629

### 2_4_(2) Random Forest 사용해 보기

In [18]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
random_pred = random_forest.predict(X_test)
random_pred

array([1, 1, 0, 2, 2, 0, 0, 2, 0, 0, 1, 2, 0, 0, 0, 2, 1, 1, 0, 0, 1, 2,
       2, 2, 1, 2, 0, 0, 0, 0, 2, 0, 2, 1, 0, 1, 2, 0, 2, 2, 0, 1, 2, 2,
       0, 0, 0, 2, 0, 1, 1, 1, 1, 1])

In [19]:
print(classification_report(y_test, random_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        22
           1       1.00      0.94      0.97        16
           2       0.94      1.00      0.97        16

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



### 2_4_(3) SVM 사용해 보기

In [20]:
svm_model = svm.SVC()

print(svm_model._estimator_type)

classifier


In [21]:
svm_model.fit(X_train, y_train)
svm_pred = svm_model.predict(X_test)
svm_pred

array([1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 1, 1])

In [22]:
print(classification_report(y_test, svm_pred))

              precision    recall  f1-score   support

           0       0.95      0.86      0.90        22
           1       0.44      0.94      0.60        16
           2       0.00      0.00      0.00        16

    accuracy                           0.63        54
   macro avg       0.46      0.60      0.50        54
weighted avg       0.52      0.63      0.55        54



/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### 2_4_(4) SGD Classifier 사용해 보기

In [23]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

print(sgd_model._estimator_type)

classifier


In [24]:
sgd_model.fit(X_train, y_train)
sgd_pred = sgd_model.predict(X_test)
sgd_pred

array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1])

In [25]:
print(classification_report(y_test, sgd_pred))

              precision    recall  f1-score   support

           0       0.47      1.00      0.64        22
           1       0.86      0.38      0.52        16
           2       0.00      0.00      0.00        16

    accuracy                           0.52        54
   macro avg       0.44      0.46      0.39        54
weighted avg       0.44      0.52      0.41        54



/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### 2_4_(5) logistic Regression 사용해 보기

In [26]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

print(logistic_model._estimator_type)

classifier


In [27]:
logistic_model.fit(X_train, y_train)
logistic_pred = logistic_model.predict(X_test)
logistic_pred

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


array([1, 1, 0, 2, 2, 0, 0, 2, 0, 0, 1, 2, 0, 0, 0, 1, 1, 1, 0, 0, 1, 2,
       2, 2, 1, 2, 0, 0, 0, 0, 2, 0, 2, 1, 0, 1, 2, 0, 2, 2, 0, 1, 2, 2,
       0, 0, 0, 2, 0, 1, 1, 1, 1, 1])

In [28]:
print(classification_report(y_test, logistic_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        22
           1       0.94      0.94      0.94        16
           2       0.94      0.94      0.94        16

    accuracy                           0.96        54
   macro avg       0.96      0.96      0.96        54
weighted avg       0.96      0.96      0.96        54



## 2_5. 모델 평가

- [sklearn.metrics 평가 지표에 대한 설명](https://velog.io/@cha-suyeon/%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D-sklearn.metrics-%ED%8F%89%EA%B0%80-%EC%A7%80%ED%91%9C)

### Data info
1. Classes: 3
2. Samples per class: [59,71,48]
3. Samples total: 178
4. Dimensionality: 13
5. Features: real, positive

#### 1) 학습 방법
- Machine Learning을 이용하여 분류기(Classifier)를 만들어서 와인의 품종을 분류, 예측하였음
- 5개의 모델을 사용하였고, 학습 데이터 70%, 테스트 데이터 30%로 나누어 훈련시키고 분류기 모델을 검증하였음

#### 2) 선택한 평가 지표
- recall

#### 3) 선택 이유
- wine dataset은 3개의 포도 품종으로 만든 class가 있고 13개의 특성(feature) 값을 측정한 다중 클래스 분류 데이터셋입니다. 클래스가 3개이기 때문에, 실제 데이터셋에 대해 얼마나 잘 예측하는 지가 중요하므로 recall의 wighted avg로 평가합니다.

#### 4) 분류 결과

||model|recall-weighted avg|
|------|---|---|
|0|Decision Tree|0.96|
|1|Random Forest|0.98|
|2|Svm|0.63|
|3|SGD Classifier|0.67|
|4|Logistic Regression|0.96|

#### 5) 결과 요약
- Random Forest가 0.98로 가장 높다.
- Svm, SGD 는 0.63, 0.67로 가장 낮은 분류 결과를 보여준다.

#### 6) 결과 해석
- Random Forest는 Decision Tree의 과적합 되는 경향을 회피할 수 있는 방법이고, Decision Tree의 조합이라 볼 수 있다.
- 독립적으로 임의의 벡터 샘플에 따라 각각의 트리가 달라지며, Forest안의 모든 트리에 배분된다.
- 따라서, wine 데이터셋의 경우 특성(feature)이 13개나 되기 때문에 feature에 따른 종류가 Random Forest로 더 확실하게 구분이 되었음